10 Notes/Component Specification - Store.md

```diff
@@ -0,0 +1,470 @@
+| `getCidAndProof(treeCid: Cid, index: Natural)` | Retrieve leaf CID with inclusion proof | Tree CID, index | `Future[?!(Cid, CodexProof)]` |
+| `putBlock(blk: Block, ttl: Duration)` | Store block with quota enforcement | Block, optional TTL | `Future[?!void]` |
+| `putCidAndProof(treeCid: Cid, index: Natural, blkCid: Cid, proof: CodexProof)` | Store leaf metadata with ref counting | Tree CID, index, block CID, proof | `Future[?!void]` |
+| `hasBlock(...)` | Check block existence (CID or tree leaf) | CID / Tree CID + index | `Future[?!bool]` |
+| `delBlock(...)` | Delete block/tree leaf (with ref count checks) | CID / Tree CID + index | `Future[?!void]` |
+| `ensureExpiry(...)` | Update expiry for block/tree leaf | CID / Tree CID + index, expiry timestamp | `Future[?!void]` |
+| `listBlocks(blockType: BlockType)` | Iterate over stored blocks | Block type | `Future[?!SafeAsyncIter[Cid]]` |
+| `getBlockExpirations(maxNumber, offset)` | Retrieve block expiry metadata | Pagination params | `Future[?!SafeAsyncIter[BlockExpiration]]` |
+| `blockRefCount(cid: Cid)` | Get block reference count | CID | `Future[?!Natural]` |
+| `reserve(bytes: NBytes)` | Reserve storage quota | Bytes | `Future[?!void]` |
+| `release(bytes: NBytes)` | Release reserved quota | Bytes | `Future[?!void]` |
+| `start()` | Initialize store | — | `Future[void]` |
+| `stop()` | Gracefully shut down store | — | `Future[void]` |
+| `close()` | Close underlying datastores | — | `Future[void]` |
+
+---
+
+## 4. Functional Requirements
+
+**Atomic Block Operations**
+
+- Store, retrieve, and delete operations must be atomic
+- Support retrieval via:
+    - Direct CID
+    - Tree-based addressing (`treeCid` + index)
+    - Unified block address
+
+**Metadata Management**
+
+- Store protocol-level metadata (e.g., sales data, storage proofs, quota usage)
+- Store block-level metadata (e.g., reference counts, total block count)
+
+**Multi-Datastore Support**
+
+- Pluggable datastore interface supporting various backends
+- Typed datastore operations for metadata type safety
+- Uniform API across implementations
+
+**Lifecycle & Maintenance**
+
+- **BlockMaintainer** service for removing expired data
+- Configurable maintenance intervals (default: 10 min)
+- Batch processing (default: 1000 blocks/cycle)
+- Cooperative scheduling to avoid blocking
+- State tracking for large datasets
+
+**Transaction Rollback & Error Recovery**
+
+- Rollback for failed multi-step operations
+- Consistent state restoration after failures
+
+---
+
+## 5. Non-Functional Requirements
+
+**Performance**
+
+- Batch metadata updates
+- Efficient key lookups with configurable prefix lengths
+- Lazy iteration and streaming for large datasets
+- Support for both fast and slower storage tiers
+
+**Security**
+
+- Verify block content integrity upon retrieval
+- Enforce quotas to prevent disk exhaustion
+- Safe orphaned data cleanup
```

Comment:
Orphaned meaning the storage request is no longer active? Wasn't that planned but not yet implemented?